File: sdk-python/obsvr/device_identity.py
"""Optional client-held device signing identity (Ed25519).

The HMAC chain is keyed from the API key, so every API-key holder — the
customer, the service, anyone who has ever seen the key — can mint a complete
valid chain. That is an integrity mechanism, not non-repudiation, and
``SECURITY.md`` says so. This module is the opt-in second seal: a device
private key the operator generates and holds, whose signature over the SAME
preimage the HMAC signs gives non-repudiation against everyone who does not
hold that key. The HMAC chain is untouched — every existing chain and
verifier keeps working, unchanged, forever; device fields are additive.

What is signed: ``obsvr-device/1 \\x00 key_id \\x00 signature_payload`` —
the version label and the key id are inside the signature, so neither can be
swapped without breaking it, and the payload is byte-identical to the HMAC's
(content, order via ``prev_sig`` linkage, and the format-3 decision fields).

Key handling rules, stated because the failure modes are known:

- **This module NEVER generates a key.** Not at init, not at verify. A
  verifier that mints key material reports every genuine record as foreign
  on a fresh machine; generation is the operator's act, out of band
  (``openssl genpkey -algorithm ed25519``).
- The key id is ``sha256(raw 32-byte public key)`` hex, first 16 chars —
  derived, never configured, so a record cannot claim an identity its
  signature does not prove.
- The key id on an event is a HINT for key selection. Verification trusts
  only keys pinned out of band; an event signed by an unpinned key is
  reported as foreign, never trusted on first use.
- With a key file configured and no Ed25519 backend importable, ``init()``
  REFUSES rather than shipping unsigned events under a config that promised
  signing — a silent no-op here is the defect, not the safeguard. Python
  needs ``pip install "obsvr-sdk[crypto]"`` (or PyNaCl); TypeScript signs
  with ``node:crypto`` and needs nothing.

Twin: sdk-typescript/src/proxy/device-identity.ts. The signature bytes are
pinned cross-language in ``conformance/fixtures/signing_vectors.json``
(Ed25519 is deterministic per RFC 8032, so one literal pin covers both).
"""

import base64
import binascii
import hashlib
import os
from typing import Any, Callable, Optional

#: Leads the device preimage. A version bump here is a new label, verified
#: side by side with this one — never a silent semantic change.
DEVICE_SIG_VERSION = "obsvr-device/1"

#: Length of the derived key id (hex chars of sha256(raw public key)).
DEVICE_KEY_ID_LEN = 16


class DeviceIdentityError(ValueError):
    """A configured device identity that cannot actually sign or be read.

    Raised loudly at init / CLI-argument time: the operator asked for
    non-repudiation, so anything short of a working signer must refuse.
    """


def derive_device_key_id(raw_public_key: bytes) -> str:
    """``sha256(raw 32-byte public key)`` hex, truncated to 16 chars."""
    if len(raw_public_key) != 32:
        raise DeviceIdentityError(
            f"device public key must be 32 raw bytes, got {len(raw_public_key)}"
        )
    return hashlib.sha256(raw_public_key).hexdigest()[:DEVICE_KEY_ID_LEN]


def device_preimage(key_id: str, signature_payload: str) -> bytes:
    """The exact bytes the device key signs (see module docstring)."""
    return (
        DEVICE_SIG_VERSION.encode("utf-8")
        + b"\x00"
        + key_id.encode("utf-8")
        + b"\x00"
        + signature_payload.encode("utf-8")
    )


def _decode_key_material(text: str, *, what: str) -> Optional[bytes]:
    """32 raw bytes from a hex or base64 rendering, or None if neither."""
    stripped = "".join(text.split())
    if len(stripped) == 64:
        try:
            raw = binascii.unhexlify(stripped)
            if len(raw) == 32:
                return raw
        except (binascii.Error, ValueError):
            pass
    try:
        raw = base64.b64decode(stripped, validate=True)
        if len(raw) == 32:
            return raw
    except (binascii.Error, ValueError):
        pass
    return None


class DeviceSigner:
    """A loaded, working signing identity: ``sign()`` or it does not exist."""

    def __init__(self, sign: Callable[[bytes], bytes], raw_public_key: bytes) -> None:
        self._sign = sign
        self.raw_public_key = raw_public_key
        self.public_key_b64 = base64.b64encode(raw_public_key).decode("ascii")
        self.key_id = derive_device_key_id(raw_public_key)

    def sign_payload(self, signature_payload: str) -> str:
        """Hex Ed25519 signature over the device preimage for this payload."""
        return self._sign(device_preimage(self.key_id, signature_payload)).hex()


def _signer_from_seed(seed: bytes) -> DeviceSigner:
    """Build a signer from a raw 32-byte seed via whichever backend imports."""
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import (
            Ed25519PrivateKey,
        )

        key = Ed25519PrivateKey.from_private_bytes(seed)
        from cryptography.hazmat.primitives.serialization import (
            Encoding,
            PublicFormat,
        )

        raw_pub = key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
        return DeviceSigner(key.sign, raw_pub)
    except ImportError:
        pass

    try:
        from nacl.signing import SigningKey

        key = SigningKey(seed)
        return DeviceSigner(
            lambda message: key.sign(message).signature,
            bytes(key.verify_key),
        )
    except ImportError:
        pass

    raise DeviceIdentityError(
        "device signing is configured but no Ed25519 backend is importable; "
        'install one: pip install "obsvr-sdk[crypto]" (or pynacl)'
    )


def _signer_from_pem(pem_text: str) -> DeviceSigner:
    """Build a signer from a PKCS8 PEM (needs ``cryptography``)."""
    try:
        from cryptography.hazmat.primitives.asymmetric.ed25519 import (
            Ed25519PrivateKey,
        )
        from cryptography.hazmat.primitives.serialization import (
            load_pem_private_key,
        )
    except ImportError:
        raise DeviceIdentityError(
            "the device key file is PEM, which needs the cryptography backend; "
            'install it: pip install "obsvr-sdk[crypto]" — or store the key as '
            "a raw 32-byte seed (hex or base64), which PyNaCl can also read"
        ) from None

    try:
        key = load_pem_private_key(pem_text.encode("utf-8"), password=None)
    except Exception as exc:
        raise DeviceIdentityError(f"unreadable device key PEM: {exc}") from exc
    if not isinstance(key, Ed25519PrivateKey):
        raise DeviceIdentityError(
            f"device key must be Ed25519, got {type(key).__name__}"
        )
    from cryptography.hazmat.primitives.serialization import Encoding, PublicFormat

    raw_pub = key.public_key().public_bytes(Encoding.Raw, PublicFormat.Raw)
    return DeviceSigner(key.sign, raw_pub)


def load_device_signer(path: str) -> DeviceSigner:
    """Load the operator-generated private key at ``path``, loudly.

    Accepted forms: a PKCS8 PEM (``BEGIN PRIVATE KEY``), or a raw 32-byte
    seed rendered as 64 hex chars or base64. Anything else — including a
    missing file — raises :class:`DeviceIdentityError`. Never generates.
    """
    if not path or not os.path.isfile(path):
        raise DeviceIdentityError(
            f"device signing key file not found: {path!r} — this SDK never "
            "generates keys; create one out of band "
            "(openssl genpkey -algorithm ed25519)"
        )
    try:
        text = open(path, "r", encoding="utf-8").read()
    except OSError as exc:
        raise DeviceIdentityError(f"unreadable device key file {path!r}: {exc}") from exc

    if "BEGIN" in text:
        return _signer_from_pem(text)
    seed = _decode_key_material(text, what="private seed")
    if seed is None:
        raise DeviceIdentityError(
            f"device key file {path!r} is neither a PEM nor a 32-byte seed "
            "(hex or base64)"
        )
    return _signer_from_seed(seed)


def load_device_public_key(value: str) -> bytes:
    """Raw 32 public-key bytes from a verifier-supplied value.

    Accepts the raw key as base64 or 64-char hex, or a path to a PEM public
    key (``BEGIN PUBLIC KEY``; needs ``cryptography``). For pinning at
    verify time — the caller derives the key id, never the export.
    """
    text = value
    if os.path.isfile(value):
        try:
            text = open(value, "r", encoding="utf-8").read()
        except OSError as exc:
            raise DeviceIdentityError(
                f"unreadable device public key file {value!r}: {exc}"
            ) from exc
    if "BEGIN" in text:
        try:
            from cryptography.hazmat.primitives.asymmetric.ed25519 import (
                Ed25519PublicKey,
            )
            from cryptography.hazmat.primitives.serialization import (
                Encoding,
                PublicFormat,
                load_pem_public_key,
            )
        except ImportError:
            raise DeviceIdentityError(
                "a PEM public key needs the cryptography backend; install it "
                '(pip install "obsvr-sdk[crypto]") or supply the raw key as '
                "base64/hex"
            ) from None
        try:
            key = load_pem_public_key(text.encode("utf-8"))
        except Exception as exc:
            raise DeviceIdentityError(f"unreadable public key PEM: {exc}") from exc
        if not isinstance(key, Ed25519PublicKey):
            raise DeviceIdentityError(
                f"device public key must be Ed25519, got {type(key).__name__}"
            )
        return key.public_bytes(Encoding.Raw, PublicFormat.Raw)
    raw = _decode_key_material(text, what="public key")
    if raw is None:
        raise DeviceIdentityError(
            "device public key must be 32 raw bytes as base64 or hex, or a "
            "PEM file"
        )
    return raw


def verify_device_sig(
    raw_public_key: bytes, key_id: str, signature_payload: str, sig_hex: Any
) -> Optional[bool]:
    """True/False under the given key, or None if no backend is importable.

    ``None`` is distinct on purpose: the caller (the verifier) must report
    "could not check" as its own outcome, never fold it into valid or
    tampered.
    """
    from .policy_verify import _resolve_backend

    backend = _resolve_backend()
    if backend is None:
        return None
    if not isinstance(sig_hex, str):
        return False
    try:
        signature = binascii.unhexlify(sig_hex)
    except (binascii.Error, ValueError):
        return False
    return backend(raw_public_key, device_preimage(key_id, signature_payload), signature)

File: sdk-typescript/src/proxy/device-identity.ts
/**
 * Optional client-held device signing identity (Ed25519).
 *
 * The HMAC chain is keyed from the API key, so every API-key holder can mint
 * a complete valid chain — an integrity mechanism, not non-repudiation, and
 * SECURITY.md says so. This module is the opt-in second seal: a device
 * private key the operator generates and holds, whose signature over the
 * SAME payload the HMAC signs gives non-repudiation against everyone who
 * does not hold that key. The HMAC chain is untouched — device fields are
 * additive, so every existing chain and verifier keeps working unchanged.
 *
 * What is signed: `obsvr-device/1 \0 key_id \0 signature_payload` — the
 * version label and the key id are inside the signature, so neither can be
 * swapped without breaking it.
 *
 * Key handling rules, stated because the failure modes are known:
 * - This module NEVER generates a key — not at init, not at verify. A
 *   verifier that mints key material reports every genuine record as
 *   foreign on a fresh machine. Generation is the operator's act, out of
 *   band (`openssl genpkey -algorithm ed25519`).
 * - The key id is sha256(raw 32-byte public key) hex, first 16 chars —
 *   derived, never configured.
 * - The key id on an event is a HINT for key selection; verification trusts
 *   only keys pinned out of band. An unpinned key id is reported as
 *   foreign, never trusted on first use.
 *
 * Twin: sdk-python/obsvr/device_identity.py. Signature bytes are pinned
 * cross-language in conformance/fixtures/signing_vectors.json (Ed25519 is
 * deterministic per RFC 8032, so one literal pin covers both).
 *
 * @packageDocumentation
 */

import {
  createHash,
  createPrivateKey,
  createPublicKey,
  sign as cryptoSign,
  verify as cryptoVerify,
  type KeyObject,
} from "node:crypto";
import { existsSync, readFileSync } from "node:fs";

/** Leads the device preimage; a bump is a new label, never a silent change. */
export const DEVICE_SIG_VERSION = "obsvr-device/1";

/** Length of the derived key id (hex chars of sha256(raw public key)). */
export const DEVICE_KEY_ID_LEN = 16;

/** A configured device identity that cannot actually sign or be read. */
export class DeviceIdentityError extends Error {}

/** PKCS8 DER framing for a raw Ed25519 seed (RFC 8410). */
const PKCS8_ED25519_PREFIX = Buffer.from(
  "302e020100300506032b657004220420",
  "hex",
);
/** SPKI DER framing for a raw Ed25519 public key (RFC 8410). */
const SPKI_ED25519_PREFIX = Buffer.from("302a300506032b6570032100", "hex");

export function deriveDeviceKeyId(rawPublicKey: Buffer): string {
  if (rawPublicKey.length !== 32) {
    throw new DeviceIdentityError(
      `device public key must be 32 raw bytes, got ${rawPublicKey.length}`,
    );
  }
  return createHash("sha256")
    .update(rawPublicKey)
    .digest("hex")
    .slice(0, DEVICE_KEY_ID_LEN);
}

/** The exact bytes the device key signs. */
export function devicePreimage(keyId: string, signaturePayload: string): Buffer {
  return Buffer.concat([
    Buffer.from(DEVICE_SIG_VERSION, "utf-8"),
    Buffer.from([0]),
    Buffer.from(keyId, "utf-8"),
    Buffer.from([0]),
    Buffer.from(signaturePayload, "utf-8"),
  ]);
}

/** 32 raw bytes from a hex or base64 rendering, or null if neither. */
function decodeKeyMaterial(text: string): Buffer | null {
  const stripped = text.replace(/\s+/g, "");
  if (stripped.length === 64 && /^[0-9a-fA-F]+$/.test(stripped)) {
    const raw = Buffer.from(stripped, "hex");
    if (raw.length === 32) return raw;
  }
  if (/^[A-Za-z0-9+/]+={0,2}$/.test(stripped)) {
    const raw = Buffer.from(stripped, "base64");
    if (raw.length === 32) return raw;
  }
  return null;
}

export interface DeviceSigner {
  keyId: string;
  rawPublicKey: Buffer;
  publicKeyB64: string;
  /** Hex Ed25519 signature over the device preimage for this payload. */
  signPayload(signaturePayload: string): string;
  /** Hex Ed25519 signature over caller-supplied raw bytes (no legacy domain). */
  signBytes(message: Uint8Array): string;
}

function signerFromKeyObject(key: KeyObject): DeviceSigner {
  if (key.asymmetricKeyType !== "ed25519") {
    throw new DeviceIdentityError(
      `device key must be Ed25519, got ${String(key.asymmetricKeyType)}`,
    );
  }
  const spki = createPublicKey(key).export({ type: "spki", format: "der" });
  const rawPublicKey = Buffer.from(spki.subarray(spki.length - 32));
  const keyId = deriveDeviceKeyId(rawPublicKey);
  return {
    keyId,
    rawPublicKey,
    publicKeyB64: rawPublicKey.toString("base64"),
    signPayload: (signaturePayload: string) =>
      cryptoSign(null, devicePreimage(keyId, signaturePayload), key).toString(
        "hex",
      ),
    signBytes: (message: Uint8Array) =>
      cryptoSign(null, Buffer.from(message), key).toString("hex"),
  };
}

/**
 * Load the operator-generated private key at `path`, loudly. Accepted
 * forms: a PKCS8 PEM (`BEGIN PRIVATE KEY`) or a raw 32-byte seed as 64 hex
 * chars or base64. Anything else — including a missing file — throws.
 * Never generates.
 */
export function loadDeviceSigner(path: string): DeviceSigner {
  if (!path || !existsSync(path)) {
    throw new DeviceIdentityError(
      `device signing key file not found: ${JSON.stringify(path)} — this SDK ` +
        "never generates keys; create one out of band " +
        "(openssl genpkey -algorithm ed25519)",
    );
  }
  const text = readFileSync(path, "utf-8");
  try {
    if (text.includes("BEGIN")) {
      return signerFromKeyObject(createPrivateKey(text));
    }
    const seed = decodeKeyMaterial(text);
    if (seed === null) {
      throw new DeviceIdentityError(
        `device key file ${JSON.stringify(path)} is neither a PEM nor a ` +
          "32-byte seed (hex or base64)",
      );
    }
    return signerFromKeyObject(
      createPrivateKey({
        key: Buffer.concat([PKCS8_ED25519_PREFIX, seed]),
        format: "der",
        type: "pkcs8",
      }),
    );
  } catch (err) {
    if (err instanceof DeviceIdentityError) throw err;
    throw new DeviceIdentityError(
      `unreadable device key file ${JSON.stringify(path)}: ${String(err)}`,
    );
  }
}

/**
 * Raw 32 public-key bytes from a verifier-supplied value: base64, 64-char
 * hex, or a path to a PEM public key. For pinning at verify time — the
 * caller derives the key id, never the export.
 */
export function loadDevicePublicKey(value: string): Buffer {
  let text = value;
  if (existsSync(value)) {
    text = readFileSync(value, "utf-8");
  }
  if (text.includes("BEGIN")) {
    try {
      const key = createPublicKey(text);
      if (key.asymmetricKeyType !== "ed25519") {
        throw new DeviceIdentityError(
          `device public key must be Ed25519, got ${String(key.asymmetricKeyType)}`,
        );
      }
      const spki = key.export({ type: "spki", format: "der" });
      return Buffer.from(spki.subarray(spki.length - 32));
    } catch (err) {
      if (err instanceof DeviceIdentityError) throw err;
      throw new DeviceIdentityError(`unreadable public key PEM: ${String(err)}`);
    }
  }
  const raw = decodeKeyMaterial(text);
  if (raw === null) {
    throw new DeviceIdentityError(
      "device public key must be 32 raw bytes as base64 or hex, or a PEM file",
    );
  }
  return raw;
}

/** True/false under the given key. Node always has an Ed25519 backend. */
export function verifyDeviceSig(
  rawPublicKey: Buffer,
  keyId: string,
  signaturePayload: string,
  sigHex: unknown,
): boolean {
  if (typeof sigHex !== "string" || !/^[0-9a-fA-F]+$/.test(sigHex)) {
    return false;
  }
  let key: KeyObject;
  try {
    key = createPublicKey({
      key: Buffer.concat([SPKI_ED25519_PREFIX, rawPublicKey]),
      format: "der",
      type: "spki",
    });
  } catch {
    return false;
  }
  return cryptoVerify(
    null,
    devicePreimage(keyId, signaturePayload),
    key,
    Buffer.from(sigHex, "hex"),
  );
}
